File: public/docs/_examples/lifecycle-hooks/ts/app/counter.component.ts
// #docregion
import {
  Component, Input, Output,
  OnChanges, SimpleChange,
} from 'angular2/core';

import {Spy} from './spy.directive';
import {LoggerService}  from './logger.service';

@Component({
  selector: 'my-counter',
  template: `
  <div class="counter">
    Counter = {{counter}}

    <h5>-- Counter Change Log --</h5>
    <div *ngFor="#chg of changeLog" my-spy>{{chg}}</div>
  </div>
  `,
  styles: ['.counter {background: LightYellow; padding: 8px; margin-top: 8px}'],
  directives:[Spy]
})
export class MyCounter implements OnChanges {
  @Input() counter: number;
  changeLog:string[] = [];

  ngOnChanges(changes: {[propertyName: string]: SimpleChange}) {

    // Empty the changeLog whenever counter goes to zero
    // hint: this is a way to respond programmatically to external value changes.
    if (this.counter === 0) {
      this.changeLog.length = 0;
    }

    // A change to `counter` is the only change we care about
    let prop = changes['counter'];
    let cur = prop.currentValue;
    let prev = JSON.stringify(prop.previousValue); // first time is {}; after is integer
    this.changeLog.push(`counter: currentValue = ${cur}, previousValue = ${prev}`);
  }

}

/***************************************/

@Component({
  selector: 'counter-parent',
  template: `
   <div class="parent">
    <h2>Counter Spy</h2>

    <button (click)="updateCounter()">Update counter</button>
    <button (click)="reset()">Reset Counter</button>

    <my-counter [counter]="value"></my-counter>

    <h4>-- Spy Lifecycle Hook Log --</h4>
    <div *ngFor="#msg of spyLog">{{msg}}</div>
   </div>
  `,
  styles: ['.parent {background: gold;}'],
  directives: [MyCounter],
  providers: [LoggerService]
})
export class CounterParentComponent {
  value: number;
  spyLog:string[] = [];

  private _logger:LoggerService;

  constructor(logger:LoggerService){
    this._logger = logger;
    this.spyLog = logger.logs;
    this.reset();
  }

  updateCounter() {
    this.value += 1;
    this._logger.tick();
  }

  reset(){
    this._logger.log('-- reset --');
    this.value=0;
    this._logger.tick();
  }
}

File: public/docs/_examples/lifecycle-hooks/dart/lib/counter_component.dart
// #docregion
import 'package:angular2/core.dart';

import 'logger_service.dart';
import 'spy_directive.dart';

@Component(
    selector: 'my-counter',
    template: '''
    <div class="counter">
      Counter = {{counter}}

      <h5>-- Counter Change Log --</h5>
      <div *ngFor="let chg of changeLog" mySpy>{{chg}}</div>
    </div>
    ''',
    styles: const [
      '.counter {background: LightYellow; padding: 8px; margin-top: 8px}'
    ],
    directives: const [SpyDirective])
class MyCounterComponent implements OnChanges {
  @Input() num counter;
  List<String> changeLog = [];

  ngOnChanges(Map<String, SimpleChange> changes) {
    // Empty the changeLog whenever counter goes to zero
    // hint: this is a way to respond programmatically to external value changes.
    if (this.counter == 0) {
      changeLog.clear();
    }

    // A change to `counter` is the only change we care about
    SimpleChange chng = changes['counter'];
    var cur = chng.currentValue;
    var prev = chng.isFirstChange() ? "{}" : chng.previousValue;
    changeLog.add('counter: currentValue = $cur, previousValue = $prev');
  }
}

@Component(
    selector: 'counter-parent',
    template: '''
    <div class="parent">
      <h2>Counter Spy</h2>

      <button (click)="updateCounter()">Update counter</button>
      <button (click)="reset()">Reset Counter</button>

      <my-counter [counter]="value"></my-counter>

      <h4>-- Spy Lifecycle Hook Log --</h4>
      <div *ngFor="let msg of logs">{{msg}}</div>
    </div>
    ''',
    styles: const ['.parent {background: gold;}'],
    directives: const [MyCounterComponent],
    providers: const [LoggerService])
class CounterParentComponent {
  final LoggerService _logger;
  num value;

  CounterParentComponent(this._logger) {
    reset();
  }

  List<String> get logs => _logger.logs;

  void updateCounter() {
    value += 1;
    _logger.tick();
  }

  void reset() {
    _logger.log('-- reset --');
    value = 0;
    _logger.tick();
  }
}
